src/lib.rs

```diff
@@ -1,215 +1,64 @@
-extern crate futures;
-extern crate parking_lot;
-
-use parking_lot::Mutex;
-use futures::prelude::*;
-use futures::task::{self, Task, AtomicTask};
-use futures::executor::{self, Notify};
-
-use std::collections::HashMap;
 use std::sync::Arc;
-use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::pin::Pin;
+use std::task::{Poll, Context};
+use std::ops::DerefMut;
+use futures::{Future, FutureExt, channel::oneshot, future::{select, Either}, executor::block_on};
+use parking_lot::Mutex;
 
-/// Future that resolves when inner work finishes or on exit signal firing.
+/// Future that resolves when the exit signal has fired.
 #[derive(Clone)]
-pub struct UntilExit<F> {
-    inner: F,
-    exit: Exit,
-}
-
-impl<F: Future> Future for UntilExit<F> {
-    type Item = Option<F::Item>;
-    type Error = F::Error;
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        match self.inner.poll() {
-            Ok(Async::Ready(x)) => Ok(Async::Ready(Some(x))),
-            Ok(Async::NotReady) => Ok(self.exit.check().map(|()| None)),
-            Err(e) => Err(e),
-        }
-    }
-}
-
-struct Notifier {
-    signalled: AtomicBool,
-    outer_task: AtomicTask,
-}
+pub struct Exit(Arc<Mutex<oneshot::Receiver<()>>>);
```

Comment:
Wait since `Exit` is cloneable and we only have `oneshot` channel it means that only one of the clones will actually get the successful response. I think you rather need `mpmc` kind of channel (fan out) here, which seems to be implemented manually in the previous code.

Reply:
Yes, only one of the clones receives `()`, but the others receive `Cancelled`, and we can use that as the same thing, so I'm not sure we actually need mpmc.

Reply:
I've pushed some tests that show that this seems to work ok.

Reply:
Hmm, I thought you will get `Cancelled` only if all senders are dropped. Are you sure it's ok to `poll` again (a non-fused `Future`) if it has completed already?

Reply:
> Yes, only one of the clones receives `()`, but the others receive `Cancelled`, and we can use that as the same thing, so I'm not sure we actually need mpmc.

That doesn't work. When you poll a `Future`, you pass a waker. Only the latest waker needs to be waken up when the `Future` is ready.
If multiple tasks poll the same future, only the last task will be woken up.

There's however this which you can use: https://docs.rs/futures/0.3.1/futures/future/trait.FutureExt.html#method.shared

Reply:
These are going to be polled from a lot of different threads and probably pretty often; most of the old design was about avoiding mutexes. This can be done asynchronously

Reply:
@tomaka cool, that works!